File: bundles/org.eclipse.rap.rwt/js/org/eclipse/rap/rwt/widgets/CoolItem.js
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

/**
 * The parameter orientation must be one of 
 * qx.constant.Layout.ORIENTATION_VERTICAL or 
 * qx.constant.Layout.ORIENTATION_HORIZONTAL
 * Note that updateHandleBounds must be called after each size manipulation.
 */
qx.OO.defineClass( 
  "org.eclipse.rap.rwt.widgets.CoolItem", 
  qx.ui.layout.CanvasLayout,
  function( orientation ) {
    qx.ui.layout.CanvasLayout.call( this );
    this._orientation = orientation;
    this._handleSize = 8;
    this._locked = false;
    // Create handle to drag this CoolItem around
    this._handle = new qx.ui.basic.Terminator();
    this._handle.setAppearance( "toolbar-part-handle-line" );
    this._handle.addEventListener( "mousedown", this._onHandleMouseDown, this );
    this._handle.addEventListener( "mousemove", this._onHandleMouseMove, this );
    this._handle.addEventListener( "mouseup", this._onHandleMouseUp, this );
    this.add( this._handle );
    // buffers zIndex during drag to be restored when dropped
    this._bufferedZIndex = null;
  }
);

/**
 * The size of the 'handle' with which the user can drag around the CoolItem.
 * Depending on the orientation that was passed to the constructor this
 * either denotes the height (vertical orientation) or width (horizontal
 * orientation) of the handle.
 * Note that updateHandleBounds must be called afterwards to reflect the change
 */
qx.Proto.setHandleSize = function( handleSize ) {
  this._handleSize = handleSize;
}

qx.Proto.getHandleSize = function() {
  return this._handleSize;
}

qx.Proto.setLocked = function( value ) {
  this._locked = value;
}

/** Updates the size and position of the handle. */
qx.Proto.updateHandleBounds = function() {
  // parameter order for setSpace: x, width, y, height
  if( this._orientation == qx.constant.Layout.ORIENTATION_VERTICAL ) {
    this._handle.setSpace( 0, this.getWidth(), 0, this._handleSize );
  } else {
    this._handle.setSpace( 0, this._handleSize, 0, this.getHeight() );
  }
}

/** React on mouseDown events on _handle widget. */
qx.Proto._onHandleMouseDown = function( evt ) {
  if( !this._locked ) {
    this._handle.setCapture( true );
    this._offsetX = evt.getPageX() - this.getLeft();
    this._offsetY = evt.getPageY() - this.getTop();
    this._bufferedZIndex = this.getZIndex();
    if( typeof( this.getZIndex() ) == "number" ) {
      this.setZIndex( this.getZIndex() + 99 );
    } else {
      this.setZIndex( 99 );
    }
  }
}

/** React on mouseMove events on _handle widget. */
qx.Proto._onHandleMouseMove = function( evt ) {
  if( !this._locked && this._handle.getCapture() ) {
    this.setLeft( evt.getPageX() - this._offsetX );
    this.setTop( evt.getPageY() - this._offsetY );
  }
}

/** React on mouseUp events on _handle widget. */
qx.Proto._onHandleMouseUp = function( evt ) {
  if( !this._locked ) {
    this._handle.setCapture( false );
    this.setZIndex( this._bufferedZIndex );
    // Send request that informs about dragged CoolItem
    if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
      var widgetManager = org.eclipse.rap.rwt.WidgetManager.getInstance();
      var id = widgetManager.findIdByWidget( this );
      var req = org.eclipse.rap.rwt.Request.getInstance();
      req.addParameter( "org.eclipse.rap.rwt.events.widgetMoved", id );
      req.addParameter( id + ".bounds.x", this.getLeft() );
      req.addParameter( id + ".bounds.y", this.getTop() );
      req.send();
    }
  }
}

qx.Proto.dispose = function() {
  if( this.getDisposed() ) {
    return true;
  }
  this._handle.removeEventListener( "mousedown", this._onHandleMouseDown, this );
  this._handle.removeEventListener( "mousemove", this._onHandleMouseMove, this );
  this._handle.removeEventListener( "mouseup", this._onHandleMouseUp, this );
  if( this._handle != null ) {
    this._handle.dispose();
  }
  return qx.ui.layout.CanvasLayout.prototype.dispose.call( this );
}

File: bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/widgets/CoolItem.java
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

package org.eclipse.rap.rwt.widgets;

import java.util.Arrays;
import java.util.Comparator;
import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.graphics.Point;
import org.eclipse.rap.rwt.graphics.Rectangle;
import org.eclipse.rap.rwt.internal.widgets.ItemHolder;



/**
 * TODO [rh] JavaDoc
 */
public class CoolItem extends Item {

  private static final class CoolItemOrderComparator implements Comparator {

    public int compare( final Object object1, final Object object2 ) {
      int result;
      CoolItem item1 = ( CoolItem )object1;
      CoolItem item2 = ( CoolItem )object2;
      if( item1.getOrder() > item2.getOrder() ) {
        result = +1;
      } else if( item1.getOrder() < item2.getOrder() ) {
        result = -1;
      } else {
        result = 0;
      }
      return result;
    }
  }

  // Keep in sync with defaultValue for handleSize in CoolItem.js
  public static final int HANDLE_SIZE = 3; 
  
  private final CoolBar parent;
  private int order;
  // TODO [rh] reasonable default value
  private Point size = new Point( 0, 0 );
  private Control control;

  // TODO [rh] constructor missing: CoolItem(CoolBar,int,int)
  public CoolItem( final CoolBar parent, final int style ) {
    super( parent, style );
    this.parent = parent;
    ItemHolder.addItem( parent, this );
    order = parent.getItemCount() - 1;
  }
  
  ///////////////////
  // Widget overrides 
  
  public Display getDisplay() {
    checkWidget();
    return parent.getDisplay();
  }
  
  ////////////////
  // Getter/setter
  
  public CoolBar getParent () {
    checkWidget();
    return parent;
  }

  public void setControl( final Control control ) {
    checkWidget();
    if( control != null ) {
      if( control.isDisposed() ) {
        RWT.error( RWT.ERROR_INVALID_ARGUMENT );
      }
      if( control.getParent() != getParent() ) {
        RWT.error( RWT.ERROR_INVALID_PARENT );
      }
    }
    this.control = control;
  }

  public Control getControl() {
    checkWidget();
    return control;
  }
  
  ///////////////////////
  // Size-related methods
  
  public Point computeSize( final int wHint, final int hHint ) {
    checkWidget();
    Point result;
    int index = parent.indexOf( this );
    if( index == -1 ) {
      result = new Point( 0, 0 );
    } else {
      int width = wHint;
      int height = hHint;
      if( wHint == RWT.DEFAULT ) {
        width = 32;
      }
      if( hHint == RWT.DEFAULT ) {
        height = 32;
      }
      if( ( parent.style & RWT.VERTICAL ) != 0 ) {
        height += parent.getMargin( index );
      } else {
        width += parent.getMargin( index );
      }
      result = new Point( width, height );
    }
    return result;
  }

  public void setPreferredSize( final Point preferredSize ) {
    checkWidget();
    if( preferredSize == null ) {
      RWT.error( RWT.ERROR_NULL_ARGUMENT );
    }
    setPreferredSize( preferredSize.x, preferredSize.y );
  }

  public void setPreferredSize( final int wHint, final int hHint ) {
    checkWidget();
    if( parent.indexOf( this ) != -1 ) {
      int width = Math.max( 0, wHint );
      int height = Math.max( 0, hHint );
      int x, y;
      if( ( parent.style & RWT.VERTICAL ) != 0 ) {
        x = Math.max( 0, height - parent.getMargin( parent.indexOf( this ) ) );
        y = width;
      } else {
        x = Math.max( 0, width - parent.getMargin( parent.indexOf( this ) ) );
        y = height;
      }
      setSize( x, y );
    }
  }
  
  public void setSize( final int wHint, final int hHint ) {
    checkWidget();
    if( parent.indexOf( this ) != -1 ) {
      int width = Math.max( 0, wHint );
      int height = Math.max( 0, hHint );
      int x;
      int y;
      if( ( parent.style & RWT.VERTICAL ) != 0 ) {
        x = height + HANDLE_SIZE;
        y = width;
      } else {
        x = width + HANDLE_SIZE;
        y = height;
      }
      size = new Point( x, y );
    }
  }
  
  public void setSize( final Point size ) {
    checkWidget();
    if( size == null ) {
      RWT.error( RWT.ERROR_NULL_ARGUMENT );
    }
    setSize( size.x, size.y );
  }
  
  public Point getSize() {
    checkWidget();
    return new Point( size.x, size.y );
  }
  
  public Rectangle getBounds() {
    checkWidget();
    int left = 0;
    int top = 0;
    CoolItem[] items = getOrderedItems();
    for( int i = 0; items[ i ] != this && i < items.length; i++ ) {
      Point itemSize = parent.getItem( i ).getSize();
      if( ( parent.style & RWT.VERTICAL ) != 0 ) {
        top += itemSize.x;
      } else {
        left += itemSize.x;
      }
    }
    return new Rectangle( left, top, size.x, size.y );
  }

  ////////////////////////////
  // Item overrides - disposal 
  
  protected void releaseChildren() {
  }

  protected void releaseParent() {
    ItemHolder.removeItem( parent, this );
  }

  protected void releaseWidget() {
    control = null;
  }
  
  /////////////////////////////////////////
  // Helping methods to maintain item order
  
  int getOrder() {
    return order;
  }

  void setOrder( final int order ) {
    this.order = order;
  }

  private CoolItem[] getOrderedItems() {
    CoolItem[] result = parent.getItems();
    Arrays.sort( result, new CoolItemOrderComparator() );
    return result;
  }
}
